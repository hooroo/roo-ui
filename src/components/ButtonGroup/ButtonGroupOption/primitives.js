import styled from '@emotion/styled';
import { themeGet } from 'styled-system';
import { hideVisually } from 'polished';

const white = themeGet('colors.white');
const blue = themeGet('colors.blue');
const dusty = themeGet('colors.greys.dusty');
const steel = themeGet('colors.greys.steel');
const charcoal = themeGet('colors.greys.charcoal');
const alto = themeGet('colors.greys.alto');
const porcelain = themeGet('colors.greys.porcelain');
const borderRadius = themeGet('radii.default');
const bold = themeGet('fontWeights.bold');

const compactPadding = props =>
  `${themeGet('space.2')(props)} ${themeGet('space.4')(props)}`;
const basePadding = props =>
  `${themeGet('space.3')(props)} ${themeGet('space.5')(props)}`;

const getFontSize = ({ size = 'base' }) => themeGet(`fontSizes.${size}`);

export const HiddenInput = styled.input`
  ${hideVisually()};
`;

export const ButtonLabel = styled.label`
  position: relative;
  flex: 1 1 auto;
  width: ${props => 100 / props.totalOptions}%;
  white-space: nowrap;
  text-align: center;
  user-select: none;
  padding: ${props => (props.size === 'sm' ? compactPadding : basePadding)};
  background-color: ${porcelain};
  cursor: pointer;
  font-size: ${getFontSize};
  line-height: ${themeGet('lineHeights.normal')};
  color: ${steel};
  font-weight: ${bold};
  transition: font-size 200ms ease-in-out;

  border-left: 1px solid ${alto};
  border-right: 1px solid ${alto};
  border-top: 2px solid ${alto};
  border-bottom: 2px solid ${alto};

  &:first-of-type {
    border-left: 2px solid ${alto};
    border-top-left-radius: ${borderRadius};
    border-bottom-left-radius: ${borderRadius};
  }

  &:last-of-type {
    border-right: 2px solid ${alto};
    border-top-right-radius: ${borderRadius};
    border-bottom-right-radius: ${borderRadius};
  }

  ${HiddenInput}:checked + & {
    color: ${charcoal};
    border-color: ${blue};
    background-color: ${white};
    cursor: unset;
  }

  ${HiddenInput}:focus + & {
    border-color: ${dusty};
  }

  ${HiddenInput}:disabled + & {
    color: ${dusty};
    font-weight: ${bold};
    background-color: ${porcelain};
    border-color: ${alto};
    cursor: unset;
  }
`;
