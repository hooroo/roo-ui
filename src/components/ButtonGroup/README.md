# ButtonGroup

<!-- STORY -->

## Installation

```shell
$ yarn add roo-ui
```

## Example

```js
import { ButtonGroup } from 'roo-ui/components';

const options = [
  { label: 'One',   value: 'one' },
  { label: 'Two',   value: 'two' },
  { label: 'Three', value: 'three' },
];

const onChange = (value) => {
  ...
};

export default (
  <ButtonGroup name="group" value="two" options={options} size="sm" onChange={onChange} />
);
```

## Properties

| Name       | Description                                                    | Type       | Default | Required? |
| ---------- | -------------------------------------------------------------- | ---------- | ------- | --------- |
| `name`     | name assigned to each underlying option (radio) element        | `string`   | -       | ✔︎        |
| `value`    | default/selected option                                        | `string`   | -       | ✔︎        |
| `options`  | selectable options (each object must have `label` and `value`) | `array`    | -       | ✔︎        |
| `onChange` | fn to execute on change                                        | `function` | -       | ✔︎        |
| `disabled` | component children                                             | `boolean`  | false   |           |
| `size`     | determine font and padding sizes                               | `string`   | base    |           |

## Customization

This component can be customized with [styled-system](https://jxnblk.com/styled-system) by passing props supported by [`<Flex />`](../Box/README.md), and [`<Box />`](../Box/README.md)
