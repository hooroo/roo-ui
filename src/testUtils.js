import React from 'react';
import { mount, shallow } from 'enzyme';

export const mountWithTheme = (tree, theme) => {
  const themedTree = React.cloneElement(tree, { theme });
  return mount(themedTree);
};

export const shallowWithTheme = (tree, theme) => {
  const themedTree = React.cloneElement(tree, { theme });
  return shallow(themedTree);
};
