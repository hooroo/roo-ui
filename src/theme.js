import { rem } from 'polished';
import range from 'lodash/range';

import * as icons from './defaultIcons';

const namedColors = {
  white: '#FFFFFF',
  brightSun: '#FBCB3B',
  red: '#E00E00',
  maroon: '#870000',
  blue: '#8DE2E0',
  lightBlue: '#E7F7F7',
  orange: '#ED710B',
  lightOrange: '#FCEBCD',
  green: '#35A509',
  lightGreen: '#DEF2DE',
};

const brand = {
  primary: '#E40000',
  secondary: '#8DE2E0',
  tertiary: '#D20000',
  quaternary: '#F9F3E9',
};

const greys = {
  charcoal: '#323232',
  steel: '#666666',
  alto: '#DADADA',
  porcelain: '#F4F5F6',
  dusty: '#999999',
  mineShaft: '#222222',
};

const ui = {
  error: namedColors.orange,
  errorBackground: namedColors.lightOrange,
  info: namedColors.blue,
  infoBackground: namedColors.lightBlue,
  link: namedColors.red,
  linkHover: namedColors.maroon,
  success: namedColors.green,
  successBackground: namedColors.lightGreen,
};

const colors = {
  brand,
  greys,
  ui,
  ...namedColors,
};

const alertStyles = {
  default: { bg: 'greys.porcelain' },
  error: {
    bg: 'ui.errorBackground',
    icon: {
      name: 'warning',
      color: 'ui.error',
    },
  },
  info: {
    bg: 'ui.infoBackground',
    icon: {
      name: 'info',
      color: 'greys.charcoal',
    },
  },
  success: {
    bg: 'ui.successBackground',
    icon: {
      name: 'checkCircle',
      color: 'ui.success',
    },
  },
};

const borders = [0, '1px solid', '2px solid', '3px solid'];

const breakpoints = ['40rem', '52rem', '64rem'];

const buttons = {
  default: {
    color: 'white',
    backgroundColor: 'greys.charcoal',
  },
  primary: {
    color: 'white',
    backgroundColor: 'brand.primary',
  },
};

const fontFamily = 'Ciutadella, sans-serif';

const fontSizes = {
  xs: rem('13px'),
  sm: rem('15px'),
  base: rem('16px'),
  lg: rem('18px'),
  xl: rem('22px'),
  '2xl': rem('26px'),
  '3xl': rem('36px'),
  '4xl': rem('46px'),
};

const fontWeights = {
  normal: 500,
  bold: 600,
};

const letterSpacings = {
  normal: '0',
  wide: '0.125em',
};

const lineHeights = {
  tight: 1,
  normal: 1.35,
  loose: 1.5,
};

const maxWidths = {
  default: rem('1280px'),
};

const mediaQueries = breakpoints.map(
  n => `@media screen and (min-width: ${n})`,
);

const opacity = {
  disabled: 0.7,
};

const radii = {
  none: 0,
  default: '4px',
  rounded: '1000px',
};

const shadows = {
  none: 0,
  default: '0 1px 1px 0 rgba(0, 0, 0, 0.1)',
  heavy: '0 2px 2px 0 rgba(0, 0, 0, 0.1)',
  focus: `0 0 2px 2px ${colors.greys.dusty};`,
};

const space = range(0, 129, 4).map(i => rem(`${i}px`));

const textStyles = {
  caps: {
    textTransform: 'uppercase',
    fontWeight: fontWeights.bold,
  },

  h1: {
    fontSize: fontSizes['3xl'],
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.normal,
  },
  h2: {
    fontSize: fontSizes['2xl'],
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.normal,
  },
  h3: {
    fontSize: fontSizes.xl,
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.loose,
  },
  h4: {
    fontSize: fontSizes.lg,
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.loose,
  },
  h5: {
    fontSize: fontSizes.base,
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.loose,
  },
  h6: {
    fontSize: fontSizes.sm,
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.loose,
  },
  label: {
    fontSize: fontSizes.sm,
    lineHeight: lineHeights.loose,
    fontWeight: fontWeights.bold,
    color: colors.greys.charcoal,
  },
  text: {
    fontSize: fontSizes.base,
    lineHeight: lineHeights.loose,
    fontWeight: fontWeights.normal,
  },
};

const transitions = {
  default: '200ms ease-in',
};

export default {
  alertStyles,
  borders,
  breakpoints,
  buttons,
  colors,
  fontFamily,
  fontSizes,
  fontWeights,
  icons,
  letterSpacings,
  lineHeights,
  maxWidths,
  mediaQueries,
  opacity,
  radii,
  shadows,
  space,
  textStyles,
  transitions,
};
