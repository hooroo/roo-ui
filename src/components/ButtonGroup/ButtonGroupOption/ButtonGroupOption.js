import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { HiddenInput, ButtonLabel } from './primitives';

const ButtonGroupOption = ({
  id,
  name,
  label,
  value,
  checked,
  disabled,
  size,
  totalOptions,
  onChange,
}) => {
  const [clicked, setClicked] = useState(false);

  const handleLabelClick = () => {
    setClicked(true);
  };

  const handleInputFocus = event => {
    if (clicked) {
      setClicked(false);
      event.target.blur();
    }
  };

  const handleChange = event => {
    if (!disabled) {
      onChange(event.target.value);
    }
  };

  return (
    <>
      <HiddenInput
        type="radio"
        id={id}
        name={name}
        value={value}
        checked={checked}
        onChange={handleChange}
        onFocus={handleInputFocus}
        disabled={disabled}
      />
      <ButtonLabel
        htmlFor={id}
        onClick={handleLabelClick}
        size={size}
        totalOptions={totalOptions}
      >
        {label}
      </ButtonLabel>
    </>
  );
};

ButtonGroupOption.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  label: PropTypes.node.isRequired,
  value: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  size: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.array,
  ]),
  totalOptions: PropTypes.number.isRequired,
};

ButtonGroupOption.defaultProps = {
  checked: false,
  disabled: false,
  size: 'base',
};

export default ButtonGroupOption;
