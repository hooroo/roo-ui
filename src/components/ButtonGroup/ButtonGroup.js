import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import Flex from '../Flex';
import ButtonGroupOption from './ButtonGroupOption';

// Setting a transform here creates a stacking context, so the we can use z-index
// in ButtonGroupOption without worrying about z-index conflicts with other components
const ButtonGroupWrapper = styled(Flex)`
  transform: scale(1);
`;

const ButtonGroup = ({
  name,
  value,
  options,
  disabled,
  size,
  onChange,
  ...rest
}) => {
  return (
    <ButtonGroupWrapper role="radiogroup" aria-label={name} {...rest}>
      {options.map((option, index) => (
        <ButtonGroupOption
          key={index}
          id={`buttonGroup_${name}_${index}`}
          name={name}
          label={option.label}
          value={option.value}
          checked={option.value === value}
          disabled={disabled}
          size={size}
          totalOptions={options.length}
          onChange={onChange}
        />
      ))}
    </ButtonGroupWrapper>
  );
};

ButtonGroup.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.node.isRequired,
      value: PropTypes.string.isRequired,
    }),
  ).isRequired,
  onChange: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  size: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.array,
  ]),
};

ButtonGroup.defaultProps = {
  disabled: false,
  size: 'base',
};

export default ButtonGroup;
