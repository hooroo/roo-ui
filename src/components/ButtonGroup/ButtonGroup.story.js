import React from 'react';
import { withDocs } from 'storybook-readme';
import noop from 'lodash/noop';

import ButtonGroup from './';
import README from './README.md';

const DEFAULT_OPTIONS = [
  { label: 'Upcoming', value: 'upcoming' },
  { label: 'Past', value: 'past' },
  { label: 'Cancelled', value: 'cancelled' },
];

export default {
  title: 'Components/ButtonGroup',
  decorators: [withDocs(README)],
};

export const Default = () => (
  <ButtonGroup
    name="bookings-by-type"
    value="upcoming"
    options={DEFAULT_OPTIONS}
    onChange={noop}
  />
);

Default.story = {
  name: 'default',
};

export const Disabled = () => (
  <ButtonGroup
    name="bookings-by-type"
    value="cancelled"
    disabled={true}
    options={DEFAULT_OPTIONS}
    onChange={noop}
  />
);

Disabled.story = {
  name: 'disabled',
};

export const TwoOptions = () => (
  <ButtonGroup
    name="two-options"
    value="cash"
    options={[
      { label: 'Use cash', value: 'cash' },
      { label: 'Use points', value: 'points' },
    ]}
    onChange={noop}
  />
);

TwoOptions.story = {
  name: 'two-options',
};
