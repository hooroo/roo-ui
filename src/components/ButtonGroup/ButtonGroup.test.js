import React from 'react';
import theme from 'theme';
import { shallowWithTheme } from 'testUtils';
import { axe } from 'jest-axe';
import ButtonGroup from './';
import ButtonGroupOption from './ButtonGroupOption';

describe('<ButtonGroup />', () => {
  let props;
  let optionNodes;
  let wrapper;

  const render = () => shallowWithTheme(<ButtonGroup {...props} />, theme);

  beforeEach(() => {
    props = {
      name: 'testButtonGroup',
      value: 'one',
      disabled: false,
      options: [
        {
          label: 'One',
          value: 'one',
        },
        {
          label: 'Two',
          value: 'two',
        },
      ],
      onChange: jest.fn(),
      size: 'sm',
    };

    wrapper = render();
    optionNodes = wrapper.find(ButtonGroupOption);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has no accessibility errors', async () => {
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('passes label prop to each ButtonGroupOption', () => {
    const values = optionNodes.map(node => node.prop('label'));
    expect(values).toEqual(['One', 'Two']);
  });

  it('passes name prop to each ButtonGroupOption', () => {
    const values = optionNodes.map(node => node.prop('name'));
    expect(values).toEqual([props.name, props.name]);
  });

  it('passes value prop to each ButtonGroupOption', () => {
    const values = optionNodes.map(node => node.prop('value'));
    expect(values).toEqual(['one', 'two']);
  });

  it('passes checked prop to each ButtonGroupOption', () => {
    const values = optionNodes.map(node => node.prop('checked'));
    expect(values).toEqual([true, false]);
  });

  it('passes disabled prop to each ButtonGroupOption', () => {
    const values = optionNodes.map(node => node.prop('disabled'));
    expect(values).toEqual([false, false]);
  });

  it('passes size prop to each ButtonGroupOption', () => {
    const values = optionNodes.map(node => node.prop('size'));
    expect(values).toEqual(['sm', 'sm']);
  });

  it('passes id prop to each ButtonGroupOption', () => {
    const values = optionNodes.map(node => node.prop('id'));
    expect(values).toEqual([
      'buttonGroup_testButtonGroup_0',
      'buttonGroup_testButtonGroup_1',
    ]);
  });

  describe('onChange', () => {
    let values;

    beforeEach(() => {
      values = optionNodes.map(node => node.prop('onChange'));
      values.forEach(fn => fn());
    });

    it('passes a handleChange prop to each ButtonGroupOption', () => {
      expect(values).toEqual([props.onChange, props.onChange]);
    });

    it('invokes the onChange fn passed to ButtonGroup when called', () => {
      expect(props.onChange).toHaveBeenCalledTimes(2);
    });

    describe('on value prop change', () => {
      it('updates the checked ButtonGroupOption', () => {
        wrapper.setProps({ value: 'two' });

        expect(
          wrapper.find('ButtonGroupOption[value="two"]').prop('checked'),
        ).toEqual(true);
      });
    });
  });
});
