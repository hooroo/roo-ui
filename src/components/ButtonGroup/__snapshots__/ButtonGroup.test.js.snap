// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> renders correctly 1`] = `
<ButtonGroupWrapper
  aria-label="testButtonGroup"
  role="radiogroup"
>
  <ButtonGroupOption
    checked={true}
    disabled={false}
    id="buttonGroup_testButtonGroup_0"
    key="0"
    label="One"
    name="testButtonGroup"
    onChange={[Function]}
    size="sm"
    totalOptions={2}
    value="one"
  />
  <ButtonGroupOption
    checked={false}
    disabled={false}
    id="buttonGroup_testButtonGroup_1"
    key="1"
    label="Two"
    name="testButtonGroup"
    onChange={[Function]}
    size="sm"
    totalOptions={2}
    value="two"
  />
</ButtonGroupWrapper>
`;
